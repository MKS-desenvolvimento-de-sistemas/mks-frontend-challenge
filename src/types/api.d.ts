export interface IProduct {
  id: number;
  name: string;
  brand: string;
  description: string;
  price: string;
  photo: string;
  createdAt: string;
  updatedAt: string;
}

export interface IProductCart extends IProduct {
  quantity: number;
}

export interface IProducts {
  products: IProduct[];
}
