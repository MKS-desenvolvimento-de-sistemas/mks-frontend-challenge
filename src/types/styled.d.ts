import {} from "styled-components";
import { theme } from "../styles/themes/";

declare module "styled-components" {
  export interface DefaultTheme extends theme {}
}
