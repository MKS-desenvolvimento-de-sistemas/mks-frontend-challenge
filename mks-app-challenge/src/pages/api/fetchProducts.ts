import { DataApi, Product } from "@/@types/Products";
import axios from "axios";

const BASE_URL = 'https://mks-frontend-challenge-04811e8151e6.herokuapp.com/api/v1';

export const fetchProducts = async (page: number, rows: number): Promise<Product> => {
  try {
    const response = await axios.get(`${BASE_URL}/products`, {
      params: {
        page,
        rows,
        sortBy: 'id',
        orderBy: 'DESC',
      },
    });

    console.log(response.data.products);


    return response.data.products;
  } catch (error) {

    throw new Error('Erro ao buscar produtos da API');
  }
};