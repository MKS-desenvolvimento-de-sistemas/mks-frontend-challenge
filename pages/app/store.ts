import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import counterReducer from './counter/counterSlice';
import productsReducer from './state/actions/productsReducers';


export const store = configureStore({
  reducer: {
    counter: counterReducer,
    products: productsReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
