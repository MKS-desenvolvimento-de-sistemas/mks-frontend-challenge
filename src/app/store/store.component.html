<div class="store">
    <h1>Store</h1>
    <div *ngIf="isLoading">Loading...</div>
    <app-product-card *ngFor="let product of products" [product]="product" (addToCartEvent)="addToCart($event)"></app-product-card>
  </div>
  