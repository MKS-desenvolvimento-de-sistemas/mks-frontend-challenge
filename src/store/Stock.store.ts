import {createSlice} from '@reduxjs/toolkit';

const carrinho = createSlice({
	name: 'carrinho',
	initialState: {
		counter: 0,
		cartOpen: 'none',
	},
	reducers: {
		increment(state) {
			state.counter += 1;
		},
		decrement(state) {
			state.counter -= 1;
		},
		abrirCarrinho(state) {
			state.cartOpen = 'flex';
		},
		fecharCarrinho(state) {
			state.cartOpen = 'none';
		},
	},
});

export const {
	decrement,
	increment,
	abrirCarrinho,
	fecharCarrinho,
} = carrinho.actions;
export default carrinho.reducer;
