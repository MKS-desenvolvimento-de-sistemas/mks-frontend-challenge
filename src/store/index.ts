import { createSlice } from "@reduxjs/toolkit";
import { configureStore } from "@reduxjs/toolkit";
import { ICartItem } from "../pages/Cart";
const initialState = { items: [] } as any;

const cartSlice = createSlice({
  name: "carrinho",
  initialState,
  reducers: {
    addToCart(state, action) {
      state.items.push(action.payload);
    },

    addAdditional(state, action) {
      if (!state.items.some((item) => item.id === action.payload.id)) return;
      const item = state.items.find((item) => item.id === action.payload.id);

      state.items.push(item);
    },
    decreaseItem(state, action) {
      const remainingItems = state.items.filter(
        (item) => item.id === action.payload.id
      );

      if (remainingItems.length === 1) return;

      const itemIndex = state.items.findIndex(
        (el) => el.id === action.payload.id
      );

      state.items.splice(itemIndex, 1);
    },
    deleteItem(state, action) {
      const arr = state.items.filter((item) => item.id !== action.payload.id);
      state.items = arr;
      // do {
      //   const itemIndex = state.items.findIndex(
      //     (el) => el.id === action.payload.id
      //   );

      //   state.items.splice(itemIndex, 1);
      // } while (state.items.some((item) => item.id === action.payload.id));
    },
    deleteAllItems(state) {
      state.items = [];
    },
  },
});

const store = configureStore({
  reducer: cartSlice.reducer,
});

export const cartActions = cartSlice.actions;

export default store;
