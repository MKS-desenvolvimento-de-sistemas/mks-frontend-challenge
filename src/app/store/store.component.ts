import { Component, OnInit } from '@angular/core';
import { ApiService, Product } from '..//services/api.service';

@Component({
  selector: 'app-store',
  templateUrl: './store.component.html',
  styleUrls: ['./store.component.scss']
})
export class StoreComponent implements OnInit {
  products: Product[] = [];
  isLoading = true;

  constructor(private apiService: ApiService) {}

  ngOnInit() {
    this.apiService.fetchProducts().then(products => {
      this.products = products;
      this.isLoading = false;
    });
  }

  addToCart(product: Product) {
    // Implementar a l√≥gica de adicionar ao carrinho
    console.log('Adicionar ao carrinho:', product);
  }
}
