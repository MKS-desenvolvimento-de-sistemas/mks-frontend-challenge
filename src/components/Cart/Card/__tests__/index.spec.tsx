import "@testing-library/jest-dom";
import {
  fireEvent,
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import CartCard from "../";
import renderer from "react-test-renderer";
import { IProductCart } from "../../../../types/api";
import { Provider } from "react-redux";
import store from "../../../../redux/";

const productExample: IProductCart = {
  id: 1,
  name: "EXEMPLO 1",
  brand: "Exemplo 1",
  description: "Exemplo 1",
  price: "200",
  photo: "/vercel.svg",
  updatedAt: "2022",
  createdAt: "2022",
  quantity: 2,
};

const formatterDescription =
  productExample.description.length > 63
    ? productExample.description.slice(0, 64) + "..."
    : productExample.description;

describe("CartCard Component", () => {
  it("should render the cart card correctly", () => {
    render(
      <Provider store={store}>
        <CartCard {...productExample} />
      </Provider>
    );

    const card = screen.getByTestId("cart-card");
    const cardImage = screen.getByTestId("cart-card--image");

    expect(cardImage).toHaveStyle({
      backgroundImage: `url(${productExample.photo})`,
    });

    expect(card).toHaveTextContent(productExample.name);
    expect(card).toHaveTextContent(`${productExample.quantity}`);
    expect(card).toHaveTextContent(`R$${productExample.price}`);
    expect(card).toHaveTextContent("+");
    expect(card).toHaveTextContent("-");
  });

  it("should update the quantity of the product in the cart", () => {
    render(
      <Provider store={store}>
        <CartCard {...productExample} />
      </Provider>
    );

    const card = screen.getByTestId("cart-card");
    const decrementButton = screen.getByTestId("decrement-button");
    const incrementButton = screen.getByTestId("increment-button");

    // INICIA EM 2
    expect(card).toHaveTextContent("2");

    // DIMINUI UMA UNIDADE
    fireEvent.click(decrementButton);

    expect(card).toHaveTextContent("1");

    // NÃƒO DIMINUI
    fireEvent.click(decrementButton);

    expect(card).toHaveTextContent("1");

    // AUMENTA UMA UNIDADE
    fireEvent.click(incrementButton);

    expect(card).toHaveTextContent("2");
  });

  it("should remove product from cart", async () => {
    render(
      <Provider store={store}>
        <CartCard {...productExample} />
      </Provider>
    );

    const removeButton = screen.getByTestId("cart-card--remove");

    fireEvent.click(removeButton);

    const card = await screen.findByTestId("cart-card");

    waitForElementToBeRemoved(card).then(() => {
      expect(card).not.toBeInTheDocument();
    });
  });

  it("should match the snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <CartCard {...productExample} />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
