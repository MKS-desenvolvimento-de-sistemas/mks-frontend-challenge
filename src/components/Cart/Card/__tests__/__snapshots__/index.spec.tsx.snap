// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartCard Component should match the snapshot 1`] = `
<li
  className="sc-eCYdqJ copgrZ"
  data-testid="cart-card"
>
  <button
    className="sc-jSMfEi dPaeKE"
    data-testid="cart-card--remove"
    onClick={[Function]}
    title="Remover do Carrinho"
    type="button"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        {
          "color": undefined,
        }
      }
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
      />
    </svg>
  </button>
  <div
    className="sc-dkzDqf sc-gKXOVf gDJlNm gkZsLx"
    data-testid="cart-card--image"
  />
  <div
    className="sc-iBkjds gHxvMP"
  >
    <h4>
      EXEMPLO 1
    </h4>
    <div
      className="sc-ftvSup kLLhow"
    >
      <small>
        Qtd:
      </small>
      <button
        className="sc-papXJ GKtMD"
        data-testid="decrement-button"
        onClick={[Function]}
        title="Diminuir Quantidade"
        type="button"
      >
        -
      </button>
      <span
        data-testid="quantity-display"
      >
        2
      </span>
      <button
        className="sc-papXJ GKtMD"
        data-testid="increment-button"
        onClick={[Function]}
        title="Aumentar Quantidade"
        type="button"
      >
        +
      </button>
    </div>
    <div
      className="sc-evZas kgTYQs"
      data-testid="price"
      type="cart"
    >
      R$200
    </div>
  </div>
</li>
`;
